File: backend/app/schemas/dish.py
"""Pydantic schemas for Dish CRUD."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.restaurant_constants import DISH_ALLERGENS


def _normalize_allergens(value: list[str] | None) -> list[str] | None:
    if value is None:
        return None
    normalized: list[str] = []
    for item in value:
        tag = item.strip().lower().replace(" ", "_").replace("-", "_")
        if tag in DISH_ALLERGENS and tag not in normalized:
            normalized.append(tag)
    return normalized


class DishCreate(BaseModel):
    restaurant_id: int
    category_id: int
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    price: Decimal = Field(..., gt=0, max_digits=10, decimal_places=2)
    cuisine: str | None = Field(None, max_length=100)
    calories: int | None = Field(None, ge=0)
    protein: Decimal | None = Field(None, ge=0)
    carbs: Decimal | None = Field(None, ge=0)
    fats: Decimal | None = Field(None, ge=0)
    fiber: Decimal | None = Field(None, ge=0)
    sugar: Decimal | None = Field(None, ge=0)
    sodium: Decimal | None = Field(None, ge=0)
    ingredients: list[str] | None = None
    allergens: list[str] | None = None
    image: str | None = Field(None, max_length=500)
    images: list[str] | None = None
    is_available: bool = True

    @field_validator("allergens")
    @classmethod
    def validate_allergens(cls, value: list[str] | None) -> list[str] | None:
        return _normalize_allergens(value)

    @field_validator("ingredients")
    @classmethod
    def validate_ingredients(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None
        return [item.strip() for item in value if item and item.strip()]


class DishUpdate(BaseModel):
    category_id: int | None = None
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    price: Decimal | None = Field(None, gt=0, max_digits=10, decimal_places=2)
    cuisine: str | None = Field(None, max_length=100)
    calories: int | None = Field(None, ge=0)
    protein: Decimal | None = Field(None, ge=0)
    carbs: Decimal | None = Field(None, ge=0)
    fats: Decimal | None = Field(None, ge=0)
    fiber: Decimal | None = Field(None, ge=0)
    sugar: Decimal | None = Field(None, ge=0)
    sodium: Decimal | None = Field(None, ge=0)
    ingredients: list[str] | None = None
    allergens: list[str] | None = None
    image: str | None = Field(None, max_length=500)
    images: list[str] | None = None
    is_available: bool | None = None

    @field_validator("allergens")
    @classmethod
    def validate_allergens(cls, value: list[str] | None) -> list[str] | None:
        return _normalize_allergens(value)

    @field_validator("ingredients")
    @classmethod
    def validate_ingredients(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None
        return [item.strip() for item in value if item and item.strip()]


class DishResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    restaurant_id: int
    category_id: int
    name: str
    description: str | None = None
    price: Decimal
    cuisine: str | None = None
    calories: int | None = None
    protein: Decimal | None = None
    carbs: Decimal | None = None
    fats: Decimal | None = None
    fiber: Decimal | None = None
    sugar: Decimal | None = None
    sodium: Decimal | None = None
    ingredients: list[str] | None = None
    allergens: list[str] | None = None
    image: str | None = None
    images: list[str] | None = None
    is_available: bool
    created_at: datetime | None = None

File: frontend/lib/models/dish.dart
import 'json_parse.dart';

/// Dish from `DishResponse` (including nested cart payloads).
class Dish {
  const Dish({
    required this.id,
    required this.name,
    required this.price,
    required this.restaurantId,
    required this.categoryId,
    this.description,
    this.cuisine,
    this.calories,
    this.protein,
    this.carbs,
    this.fats,
    this.fiber,
    this.sugar,
    this.sodium,
    this.ingredients = const [],
    this.allergens = const [],
    this.image,
    this.images = const [],
    this.isAvailable = true,
    this.createdAt,
  });

  final int id;
  final String name;
  final double price;
  final int restaurantId;
  final int categoryId;
  final String? description;
  final String? cuisine;
  final int? calories;
  final double? protein;
  final double? carbs;
  final double? fats;
  final double? fiber;
  final double? sugar;
  final double? sodium;
  final List<String> ingredients;
  final List<String> allergens;
  final String? image;
  final List<String> images;
  final bool isAvailable;
  final DateTime? createdAt;

  factory Dish.fromJson(Map<String, dynamic> json) => Dish(
        id: parseInt(json['id'], field: 'id'),
        name: parseString(json['name']),
        price: parseDouble(json['price'], field: 'price'),
        restaurantId: parseInt(json['restaurant_id'], field: 'restaurant_id'),
        categoryId: parseInt(json['category_id'], field: 'category_id'),
        description: json['description']?.toString(),
        cuisine: json['cuisine']?.toString(),
        calories: parseIntOrNull(json['calories']),
        protein: parseDoubleOrNull(json['protein']),
        carbs: parseDoubleOrNull(json['carbs']),
        fats: parseDoubleOrNull(json['fats']),
        fiber: parseDoubleOrNull(json['fiber']),
        sugar: parseDoubleOrNull(json['sugar']),
        sodium: parseDoubleOrNull(json['sodium']),
        ingredients: _stringList(json['ingredients']),
        allergens: _stringList(json['allergens']),
        image: json['image']?.toString(),
        images: _stringList(json['images']),
        isAvailable: parseBool(json['is_available']),
        createdAt: parseDateTimeOrNull(json['created_at']),
      );

  static List<String> _stringList(dynamic value) {
    if (value is! List) return const [];
    return value.map((e) => e.toString()).where((s) => s.isNotEmpty).toList();
  }

  Map<String, dynamic> toWriteJson({bool includeRestaurantId = true}) {
    return {
      if (includeRestaurantId) 'restaurant_id': restaurantId,
      'category_id': categoryId,
      'name': name,
      if (description != null) 'description': description,
      'price': price,
      if (cuisine != null && cuisine!.isNotEmpty) 'cuisine': cuisine,
      if (calories != null) 'calories': calories,
      if (protein != null) 'protein': protein,
      if (carbs != null) 'carbs': carbs,
      if (fats != null) 'fats': fats,
      if (fiber != null) 'fiber': fiber,
      if (sugar != null) 'sugar': sugar,
      if (sodium != null) 'sodium': sodium,
      if (ingredients.isNotEmpty) 'ingredients': ingredients,
      if (allergens.isNotEmpty) 'allergens': allergens,
      if (image != null && image!.isNotEmpty) 'image': image,
      if (images.isNotEmpty) 'images': images,
      'is_available': isAvailable,
    };
  }
}
